In [2]:
import pandas as pd

In [3]:
#1. Load the dataset into a Pandas DataFrame.
df = pd.read_csv("C:\\Users\\PRASHANT\\Downloads\\edtech_student_learning_data.csv")

In [3]:
#2. Display the first 10 records of the dataset.
df.head(10)

student_id student_name  gender       city enrollment_date          course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25   Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29         AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05    Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09         AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03   Generative AI   
5        1006    Student_6  Female       Pune      2024-09-18   Generative AI   
6        1007    Student_7    Male     Mumbai      2024-02-22  Data Analytics   
7        1008    Student_8    Male    Chennai      2024-07-23    Data Science   
8        1009    Student_9    Male     Mumbai      2024-05-05         AI & ML   
9        1010   Student_10  Female      Delhi      2024-08-22  Data Analytics   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  \
0    Weekday             120                108              59   
1    Weekday             120                110              57   
2    Weekday             120                 73              49   
3    Weekend             120                 71              64   
4    Weekday             120                101              41   
5    Weekend             120                 79              64   
6    Weekend             120                112              93   
7    Weekend             120                109              72   
8    Weekend             120                 83              75   
9    Weekday             120                118              70   

   assignment_score  project_score  final_rating  learning_hours  \
0                96             66           3.7               6   
1                93             83           4.5              33   
2                72             73           3.6               8   
3                91             81           2.7               5   
4                55             50           2.1              23   
5                98             57           2.6              30   
6                89             94           3.2              29   
7                77             72           4.3              14   
8                61             96           2.6              19   
9                50             89           4.2              26   

  placement_status  feedback_score  attendance_pct  salary_offered  
0           Placed               7           90.00        738963.0  
1       Not Placed               8           91.67             NaN  
2           Placed               5           60.83        501904.0  
3       Not Placed               8           59.17             NaN  
4           Placed               9           84.17        970105.0  
5           Placed               6           65.83        305730.0  
6           Placed               6           93.33        707137.0  
7           Placed               6           90.83       1053790.0  
8           Placed               8           69.17        502210.0  
9       Not Placed               8           98.33             NaN

In [4]:
#3. Display the last 5 records of the dataset.
df.tail(5)

student_id student_name  gender       city enrollment_date  \
500        1452  Student_452  Female     Mumbai      2024-03-07   
501        1276  Student_276  Female  Hyderabad      2024-10-15   
502        1210  Student_210    Male  Bangalore      2024-04-04   
503        1178  Student_178    Male      Delhi      2024-01-20   
504        1397  Student_397  Female     Mumbai      2024-05-27   

            course batch_type  total_sessions  sessions_attended  \
500  Generative AI    Weekday             120                 68   
501  Generative AI    Weekday             120                 95   
502        AI & ML    Weekday             120                 75   
503  Generative AI    Weekday             120                 80   
504  Generative AI    Weekend             120                101   

     avg_quiz_score  assignment_score  project_score  final_rating  \
500              84                84             58           2.2   
501              92                81             91           3.2   
502              40                84             77           2.4   
503              97                92             66           4.7   
504              66                83             66           3.6   

     learning_hours placement_status  feedback_score  attendance_pct  \
500              19       Not Placed               8           56.67   
501              16       Not Placed               8           79.17   
502              19           Placed               5           62.50   
503              33           Placed               5           66.67   
504              25           Placed               7           84.17   

     salary_offered  
500             NaN  
501             NaN  
502        750043.0  
503        890611.0  
504       1003022.0

In [5]:
#4. Check the shape of the dataset.
df.shape

(505, 18)

In [6]:
#5. Display all column names along with their data types.
df.dtypes

student_id             int64
student_name          object
gender                object
city                  object
enrollment_date       object
course                object
batch_type            object
total_sessions         int64
sessions_attended      int64
avg_quiz_score         int64
assignment_score       int64
project_score          int64
final_rating         float64
learning_hours         int64
placement_status      object
feedback_score         int64
attendance_pct       float64
salary_offered       float64
dtype: object

In [9]:
#6. Generate summary statistics for all numerical columns.
df.describe()                             

student_id  total_sessions  sessions_attended  avg_quiz_score  \
count   505.000000           505.0         505.000000      505.000000   
mean   1251.015842           120.0          88.102970       69.601980   
std     144.242011             0.0          17.087232       17.137226   
min    1001.000000           120.0          60.000000       40.000000   
25%    1127.000000           120.0          73.000000       56.000000   
50%    1251.000000           120.0          88.000000       69.000000   
75%    1376.000000           120.0         103.000000       84.000000   
max    1500.000000           120.0         119.000000       99.000000   

       assignment_score  project_score  final_rating  learning_hours  \
count        505.000000     505.000000    477.000000      505.000000   
mean          71.984158      75.287129      3.535220       22.128713   
std           15.777337      14.678794      0.862365        9.924180   
min           45.000000      50.000000      2.000000        5.000000   
25%           58.000000      62.000000      2.800000       14.000000   
50%           72.000000      76.000000      3.600000       23.000000   
75%           86.000000      88.000000      4.300000       31.000000   
max           99.000000      99.000000      5.000000       39.000000   

       feedback_score  attendance_pct  salary_offered  
count      505.000000      505.000000    3.360000e+02  
mean         7.041584       73.419149    7.701324e+05  
std          1.370122       14.239418    2.656396e+05  
min          5.000000       50.000000    3.005150e+05  
25%          6.000000       60.830000    5.442040e+05  
50%          7.000000       73.330000    7.788115e+05  
75%          8.000000       85.830000    1.003022e+06  
max          9.000000       99.170000    1.199092e+06

In [19]:
df.describe(include="all")

student_id student_name  gender       city enrollment_date  \
count    505.000000          505     505        505             505   
unique          NaN          500       2          6             269   
top             NaN  Student_210  Female  Hyderabad      2024-04-20   
freq            NaN            2     259        100               7   
mean    1251.015842          NaN     NaN        NaN             NaN   
std      144.242011          NaN     NaN        NaN             NaN   
min     1001.000000          NaN     NaN        NaN             NaN   
25%     1127.000000          NaN     NaN        NaN             NaN   
50%     1251.000000          NaN     NaN        NaN             NaN   
75%     1376.000000          NaN     NaN        NaN             NaN   
max     1500.000000          NaN     NaN        NaN             NaN   

                course batch_type  total_sessions  sessions_attended  \
count              505        505           505.0         505.000000   
unique               4          2             NaN                NaN   
top     Data Analytics    Weekend             NaN                NaN   
freq               143        263             NaN                NaN   
mean               NaN        NaN           120.0          88.102970   
std                NaN        NaN             0.0          17.087232   
min                NaN        NaN           120.0          60.000000   
25%                NaN        NaN           120.0          73.000000   
50%                NaN        NaN           120.0          88.000000   
75%                NaN        NaN           120.0         103.000000   
max                NaN        NaN           120.0         119.000000   

        avg_quiz_score  assignment_score  project_score  final_rating  \
count       505.000000        505.000000     505.000000    477.000000   
unique             NaN               NaN            NaN           NaN   
top                NaN               NaN            NaN           NaN   
freq               NaN               NaN            NaN           NaN   
mean         69.601980         71.984158      75.287129      3.535220   
std          17.137226         15.777337      14.678794      0.862365   
min          40.000000         45.000000      50.000000      2.000000   
25%          56.000000         58.000000      62.000000      2.800000   
50%          69.000000         72.000000      76.000000      3.600000   
75%          84.000000         86.000000      88.000000      4.300000   
max          99.000000         99.000000      99.000000      5.000000   

        learning_hours placement_status  feedback_score  attendance_pct  \
count       505.000000              505      505.000000      505.000000   
unique             NaN                2             NaN             NaN   
top                NaN           Placed             NaN             NaN   
freq               NaN              336             NaN             NaN   
mean         22.128713              NaN        7.041584       73.419149   
std           9.924180              NaN        1.370122       14.239418   
min           5.000000              NaN        5.000000       50.000000   
25%          14.000000              NaN        6.000000       60.830000   
50%          23.000000              NaN        7.000000       73.330000   
75%          31.000000              NaN        8.000000       85.830000   
max          39.000000              NaN        9.000000       99.170000   

        salary_offered  
count     3.360000e+02  
unique             NaN  
top                NaN  
freq               NaN  
mean      7.701324e+05  
std       2.656396e+05  
min       3.005150e+05  
25%       5.442040e+05  
50%       7.788115e+05  
75%       1.003022e+06  
max       1.199092e+06

## student_id

student_id have **505**value,there is no missng value
student_id mean grater than std,thats mean data is strong and stable
25% of data **1127** data lies below this value
50% of data **1251** data lies below this value
75% of data **1376** data lies below this vallue
student_id min value is 1001,and max value is 1500

## student_name

student_name have **505**value,there is no missng value
there are **500** unique value in student_name column
student_210 indicates most frequent value in that column
frequent value 2 number of time the value have been repeated

## city

city have **505**value,there is no missng value
there are **6** unique value in city column
Hyderabad indicates most frequent value in that column
frequent value **100** number of time the value have been repeated

## gender

gender have **505**value,there is no missng value
there are **2** unique value in gender column
Female indicates most frequent value in that column
frequent value **259** number of time the value have been repeated

## endrollment_date

endrollment_datehave **505**value,there is no missng value
there are **269** unique value in gender column
Data Analytics indicates most frequent value in that column
frequent value **7** number of time the value have been repeated

## course

course **505**value,there is no missng value
there are **4** unique value in gender column
2024-04-20 indicates most frequent value in that column
frequent value **143** number of time the value have been repeated

## batch_type

batch_type**2**value,there is no missng value
there are **4** unique value in gender column
weekend  indicates most frequent value in that column
frequent value **263** number of time the value have been repeated

## total_sessions

total_sessions have **505**value,there is no missng value
total_sessions mean grater than std,thats mean data is less variated
25% of data **120** data lies below this value
50% of data **120** data lies below this value
75% of data **120** data lies below this vallue
total_sessions min value is 120,and max value is 120

## sessions_attended

sessions_attended have **505**value,there is no missng value
sessions_attened mean grater than std,thats mean data is strong and stable
25% of data **73** data lies below this value
50% of data **88** data lies below this value
75% of data **103** data lies below this vallue
total_sessions min value is 60,and max value is 119

## avg_quiz_score

avg_quiz_score have **505**value,there is no missng value
avg_quiz_score mean grater than std,thats mean data is strong and stable
25% of data **56** data lies below this value
50% of data **69** data lies below this value
75% of data **84** data lies below this vallue
total_sessions min value is 40 ,and max value is 99

## assignment_score

assignment_score have **505**value,there is no missng value
assignment_score mean grater than std,thats mean data is strong and stable
25% of data **56** data lies below this value
50% of data **69** data lies below this value
75% of data **84** data lies below this vallue
total_sessions min value is 40 ,and max value is 99

## project_score

project_score have **505**value,there is no missng value
project_score mean grater than std,thats mean data is strong and stable
25% of data **62** data lies below this value
50% of data **76** data lies below this value
75% of data **88** data lies below this vallue
total_sessions min value is 50 ,and max value is 99

## final_rating

final_rating have **505**value,there is no missng value
final_rating mean grater than std,thats mean data is strong and stable
25% of data **2** data lies below this value
50% of data **3** data lies below this value
75% of data **4** data lies below this vallue
total_sessions min value is 2,and max value is 5

## learning_hours

learning_hours have **505**value,there is no missng value is present in column
learning_hours mean grater than std,thats mean data is strong and stable
25% of data **2** data lies below this value
50% of data **3** data lies below this value
75% of data **4** data lies below this vallue
total_sessions min value is 2,and max value is 5

## placement_status

city have **505**value,there is no missng value
there are **2** unique value in city column
placed indicates most frequent value in that column
frequent value **336** number of time the value have been repeated

## feedback_score

feedback_score have **505**value,there is no missng value
feedback_score mean grater than std,thats mean data is strong and stable
25% of data **6** data lies below this value
50% of data **7** data lies below this value
75% of data **8** data lies below this vallue
total_sessions min value is 5 ,and max value is 9

## attendance_pct

attendance_pct have **505**value,there is no missng value
attendance_pct mean grater than std,thats mean data is strong and stable
25% of data **60** data lies below this value
50% of data **73** data lies below this value
75% of data **85** data lies below this vallue
total_sessions min value is 50,and max value is 99

## salary_offered

salart_offered have **505**value,there is no missng value
salary_offered mean grater than std,thats mean data is strong and stable
25% of data **5** data lies below this valuew
50% of data **7** data lies below this value
75% of data **1** data lies below this vallue
total_sessions min value is 3,and max value is 1

In [23]:
#7. Identify columns that contain missing values.
df.isna().sum()

student_id             0
student_name           0
gender               505
city                   0
enrollment_date        0
course                 0
batch_type             0
total_sessions         0
sessions_attended      0
avg_quiz_score         0
assignment_score       0
project_score          0
final_rating          28
learning_hours         0
placement_status       0
feedback_score         0
attendance_pct         0
salary_offered       169
dtype: int64

In [24]:
#8. Count the number of missing values in each column.
df.isna().sum()

student_id             0
student_name           0
gender               505
city                   0
enrollment_date        0
course                 0
batch_type             0
total_sessions         0
sessions_attended      0
avg_quiz_score         0
assignment_score       0
project_score          0
final_rating          28
learning_hours         0
placement_status       0
feedback_score         0
attendance_pct         0
salary_offered       169
dtype: int64

In [40]:
#9. Calculate the percentage of missing values in the final_rating column.

In [39]:
#10. Fill missing values in the final_rating column using the mean.

In [28]:
#11. Verify that missing values in final_rating are handled.
df['final_rating'].isnull().sum()

np.int64(0)

In [29]:
#12. Check whether duplicate records exist in the dataset.
# df.isna().sum()
df[df.duplicated()]

student_id student_name gender       city enrollment_date         course  \
500        1452  Student_452    NaN     Mumbai      2024-03-07  Generative AI   
501        1276  Student_276    NaN  Hyderabad      2024-10-15  Generative AI   
502        1210  Student_210    NaN  Bangalore      2024-04-04        AI & ML   
503        1178  Student_178    NaN      Delhi      2024-01-20  Generative AI   
504        1397  Student_397    NaN     Mumbai      2024-05-27  Generative AI   

    batch_type  total_sessions  sessions_attended  avg_quiz_score  \
500    Weekday             120                 68              84   
501    Weekday             120                 95              92   
502    Weekday             120                 75              40   
503    Weekday             120                 80              97   
504    Weekend             120                101              66   

     assignment_score  project_score  final_rating  learning_hours  \
500                84             58           2.2              19   
501                81             91           3.2              16   
502                84             77           2.4              19   
503                92             66           4.7              33   
504                83             66           3.6              25   

    placement_status  feedback_score  attendance_pct  salary_offered  
500       Not Placed               8           56.67             NaN  
501       Not Placed               8           79.17             NaN  
502           Placed               5           62.50        750043.0  
503           Placed               5           66.67        890611.0  
504           Placed               7           84.17       1003022.0

In [37]:
#13. Count the total number of duplicate rows.
df.duplicated().sum()

np.int64(5)

In [31]:
#14. Display all duplicate records.
df[df.duplicated()]

student_id student_name gender       city enrollment_date         course  \
500        1452  Student_452    NaN     Mumbai      2024-03-07  Generative AI   
501        1276  Student_276    NaN  Hyderabad      2024-10-15  Generative AI   
502        1210  Student_210    NaN  Bangalore      2024-04-04        AI & ML   
503        1178  Student_178    NaN      Delhi      2024-01-20  Generative AI   
504        1397  Student_397    NaN     Mumbai      2024-05-27  Generative AI   

    batch_type  total_sessions  sessions_attended  avg_quiz_score  \
500    Weekday             120                 68              84   
501    Weekday             120                 95              92   
502    Weekday             120                 75              40   
503    Weekday             120                 80              97   
504    Weekend             120                101              66   

     assignment_score  project_score  final_rating  learning_hours  \
500                84             58           2.2              19   
501                81             91           3.2              16   
502                84             77           2.4              19   
503                92             66           4.7              33   
504                83             66           3.6              25   

    placement_status  feedback_score  attendance_pct  salary_offered  
500       Not Placed               8           56.67             NaN  
501       Not Placed               8           79.17             NaN  
502           Placed               5           62.50        750043.0  
503           Placed               5           66.67        890611.0  
504           Placed               7           84.17       1003022.0

In [26]:
#15. Remove duplicate records from the dataset.

In [40]:
df["salary_offered"].fillna(df["salary_offered"].mean(),inplace=True)

C:\Users\PRASHANT\AppData\Local\Temp\ipykernel_33996\2245153338.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["salary_offered"].fillna(df["salary_offered"].mean(),inplace=True)


In [41]:
df.isna().sum()

student_id             0
student_name           0
gender               505
city                   0
enrollment_date        0
course                 0
batch_type             0
total_sessions         0
sessions_attended      0
avg_quiz_score         0
assignment_score       0
project_score          0
final_rating           0
learning_hours         0
placement_status       0
feedback_score         0
attendance_pct         0
salary_offered         0
dtype: int64

In [39]:
#16. Verify the dataset size after removing duplicates.
print(df.size)

9090
(505, 18)


In [38]:
#17. Find the number of unique values in the city column.
df["city"].nunique()

6

In [37]:
#18. Find the number of unique values in the course column.
df["course"].nunique()

4

In [33]:
#19. Display all unique cities present in the dataset.
df["city"].unique()

array(['Hyderabad', 'Pune', 'Delhi', 'Mumbai', 'Chennai', 'Bangalore'],
      dtype=object)

In [34]:
#20. Count the number of students in each city.
df["city"].value_counts()

city
Hyderabad    100
Mumbai        95
Pune          87
Bangalore     82
Chennai       73
Delhi         68
Name: count, dtype: int64

In [35]:
#21. Count the number of students enrolled in each course.
df["course"].value_counts()

course
Data Analytics    143
Data Science      128
AI & ML           126
Generative AI     108
Name: count, dtype: int64

In [27]:
#22. Identify the course with the highest student enrollment.
df["course"].max()

'Generative AI'

In [4]:
#23. Count the number of students who are placed and not placed.
df["placement_status"].value_counts()s]

placement_status
Placed        336
Not Placed    169
Name: count, dtype: int64

In [5]:
#24. Filter students whose attendance percentage is greater than 80.
df[df.attendance_pct>=80]

student_id student_name  gender       city enrollment_date  \
0          1001    Student_1    Male  Hyderabad      2024-11-25   
1          1002    Student_2  Female  Hyderabad      2024-09-29   
4          1005    Student_5    Male     Mumbai      2024-02-03   
6          1007    Student_7    Male     Mumbai      2024-02-22   
7          1008    Student_8    Male    Chennai      2024-07-23   
..          ...          ...     ...        ...             ...   
475        1476  Student_476  Female    Chennai      2024-02-05   
477        1478  Student_478  Female     Mumbai      2024-07-06   
481        1482  Student_482  Female     Mumbai      2024-07-15   
488        1489  Student_489    Male    Chennai      2024-10-19   
504        1397  Student_397  Female     Mumbai      2024-05-27   

             course batch_type  total_sessions  sessions_attended  \
0     Generative AI    Weekday             120                108   
1           AI & ML    Weekday             120                110   
4     Generative AI    Weekday             120                101   
6    Data Analytics    Weekend             120                112   
7      Data Science    Weekend             120                109   
..              ...        ...             ...                ...   
475  Data Analytics    Weekday             120                 97   
477         AI & ML    Weekend             120                104   
481   Generative AI    Weekend             120                101   
488   Generative AI    Weekend             120                111   
504   Generative AI    Weekend             120                101   

     avg_quiz_score  assignment_score  project_score  final_rating  \
0                59                96             66           3.7   
1                57                93             83           4.5   
4                41                55             50           2.1   
6                93                89             94           3.2   
7                72                77             72           4.3   
..              ...               ...            ...           ...   
475              99                99             94           3.6   
477              88                73             87           3.8   
481              63                56             66           3.6   
488              70                49             75           3.3   
504              66                83             66           3.6   

     learning_hours placement_status  feedback_score  attendance_pct  \
0                 6           Placed               7           90.00   
1                33       Not Placed               8           91.67   
4                23           Placed               9           84.17   
6                29           Placed               6           93.33   
7                14           Placed               6           90.83   
..              ...              ...             ...             ...   
475              33           Placed               9           80.83   
477              14           Placed               5           86.67   
481              30           Placed               9           84.17   
488              26       Not Placed               7           92.50   
504              25           Placed               7           84.17   

     salary_offered  
0          738963.0  
1               NaN  
4          970105.0  
6          707137.0  
7         1053790.0  
..              ...  
475        855839.0  
477        749813.0  
481        309637.0  
488             NaN  
504       1003022.0  

[186 rows x 18 columns]

In [7]:
#25. Filter students who attended fewer than 70 sessions.
df[df.sessions_attended>=70]

student_id student_name  gender       city enrollment_date  \
0          1001    Student_1    Male  Hyderabad      2024-11-25   
1          1002    Student_2  Female  Hyderabad      2024-09-29   
2          1003    Student_3    Male       Pune      2024-12-05   
3          1004    Student_4    Male      Delhi      2024-03-09   
4          1005    Student_5    Male     Mumbai      2024-02-03   
..          ...          ...     ...        ...             ...   
499        1500  Student_500  Female       Pune      2024-06-06   
501        1276  Student_276  Female  Hyderabad      2024-10-15   
502        1210  Student_210    Male  Bangalore      2024-04-04   
503        1178  Student_178    Male      Delhi      2024-01-20   
504        1397  Student_397  Female     Mumbai      2024-05-27   

            course batch_type  total_sessions  sessions_attended  \
0    Generative AI    Weekday             120                108   
1          AI & ML    Weekday             120                110   
2     Data Science    Weekday             120                 73   
3          AI & ML    Weekend             120                 71   
4    Generative AI    Weekday             120                101   
..             ...        ...             ...                ...   
499        AI & ML    Weekday             120                 73   
501  Generative AI    Weekday             120                 95   
502        AI & ML    Weekday             120                 75   
503  Generative AI    Weekday             120                 80   
504  Generative AI    Weekend             120                101   

     avg_quiz_score  assignment_score  project_score  final_rating  \
0                59                96             66           3.7   
1                57                93             83           4.5   
2                49                72             73           3.6   
3                64                91             81           2.7   
4                41                55             50           2.1   
..              ...               ...            ...           ...   
499              57                79             78           3.8   
501              92                81             91           3.2   
502              40                84             77           2.4   
503              97                92             66           4.7   
504              66                83             66           3.6   

     learning_hours placement_status  feedback_score  attendance_pct  \
0                 6           Placed               7           90.00   
1                33       Not Placed               8           91.67   
2                 8           Placed               5           60.83   
3                 5       Not Placed               8           59.17   
4                23           Placed               9           84.17   
..              ...              ...             ...             ...   
499              21           Placed               7           60.83   
501              16       Not Placed               8           79.17   
502              19           Placed               5           62.50   
503              33           Placed               5           66.67   
504              25           Placed               7           84.17   

     salary_offered  
0          738963.0  
1               NaN  
2          501904.0  
3               NaN  
4          970105.0  
..              ...  
499       1115331.0  
501             NaN  
502        750043.0  
503        890611.0  
504       1003022.0  

[411 rows x 18 columns]

In [8]:
#26. Filter students with an average quiz score above 85.
df[df.avg_quiz_score>=85]

student_id student_name  gender       city enrollment_date  \
6          1007    Student_7    Male     Mumbai      2024-02-22   
12         1013   Student_13    Male    Chennai      2024-04-20   
14         1015   Student_15  Female      Delhi      2024-05-30   
15         1016   Student_16    Male     Mumbai      2024-07-01   
16         1017   Student_17  Female     Mumbai      2024-12-01   
..          ...          ...     ...        ...             ...   
495        1496  Student_496    Male  Hyderabad      2024-12-29   
496        1497  Student_497    Male  Hyderabad      2024-10-06   
497        1498  Student_498  Female  Bangalore      2024-03-21   
501        1276  Student_276  Female  Hyderabad      2024-10-15   
503        1178  Student_178    Male      Delhi      2024-01-20   

             course batch_type  total_sessions  sessions_attended  \
6    Data Analytics    Weekend             120                112   
12    Generative AI    Weekend             120                 84   
14   Data Analytics    Weekend             120                 88   
15   Data Analytics    Weekday             120                116   
16          AI & ML    Weekend             120                 98   
..              ...        ...             ...                ...   
495    Data Science    Weekday             120                 86   
496   Generative AI    Weekday             120                 61   
497  Data Analytics    Weekday             120                 94   
501   Generative AI    Weekday             120                 95   
503   Generative AI    Weekday             120                 80   

     avg_quiz_score  assignment_score  project_score  final_rating  \
6                93                89             94           3.2   
12               85                92             95           NaN   
14               93                71             64           3.9   
15               85                59             90           3.6   
16               91                50             68           3.4   
..              ...               ...            ...           ...   
495              94                97             57           3.1   
496              88                53             72           3.7   
497              94                53             63           3.4   
501              92                81             91           3.2   
503              97                92             66           4.7   

     learning_hours placement_status  feedback_score  attendance_pct  \
6                29           Placed               6           93.33   
12               38           Placed               7           70.00   
14               29           Placed               7           73.33   
15               32       Not Placed               7           96.67   
16                5       Not Placed               9           81.67   
..              ...              ...             ...             ...   
495              14           Placed               7           71.67   
496              28           Placed               5           50.83   
497              39           Placed               7           78.33   
501              16       Not Placed               8           79.17   
503              33           Placed               5           66.67   

     salary_offered  
6          707137.0  
12        1167965.0  
14        1128411.0  
15              NaN  
16              NaN  
..              ...  
495        572544.0  
496        834633.0  
497        333896.0  
501             NaN  
503        890611.0  

[125 rows x 18 columns]

In [9]:
#27. Filter students enrolled in the Data Science course.
df[df["course"].isin(["Data Science"])]

student_id student_name  gender       city enrollment_date        course  \
2          1003    Student_3    Male       Pune      2024-12-05  Data Science   
7          1008    Student_8    Male    Chennai      2024-07-23  Data Science   
10         1011   Student_11    Male    Chennai      2024-02-12  Data Science   
13         1014   Student_14    Male     Mumbai      2024-09-05  Data Science   
18         1019   Student_19  Female       Pune      2024-01-08  Data Science   
..          ...          ...     ...        ...             ...           ...   
487        1488  Student_488  Female       Pune      2024-09-17  Data Science   
489        1490  Student_490    Male       Pune      2024-09-12  Data Science   
490        1491  Student_491  Female  Bangalore      2024-08-09  Data Science   
495        1496  Student_496    Male  Hyderabad      2024-12-29  Data Science   
498        1499  Student_499    Male     Mumbai      2024-07-20  Data Science   

    batch_type  total_sessions  sessions_attended  avg_quiz_score  \
2      Weekday             120                 73              49   
7      Weekend             120                109              72   
10     Weekday             120                102              53   
13     Weekday             120                109              55   
18     Weekend             120                115              74   
..         ...             ...                ...             ...   
487    Weekday             120                 66              92   
489    Weekend             120                 73              75   
490    Weekday             120                 68              58   
495    Weekday             120                 86              94   
498    Weekday             120                 77              72   

     assignment_score  project_score  final_rating  learning_hours  \
2                  72             73           3.6               8   
7                  77             72           4.3              14   
10                 48             61           4.6              25   
13                 91             96           3.7              34   
18                 58             78           2.9              11   
..                ...            ...           ...             ...   
487                65             51           2.2              27   
489                92             62           4.1               6   
490                64             96           3.4              35   
495                97             57           3.1              14   
498                48             76           NaN              27   

    placement_status  feedback_score  attendance_pct  salary_offered  
2             Placed               5           60.83        501904.0  
7             Placed               6           90.83       1053790.0  
10        Not Placed               7           85.00             NaN  
13            Placed               6           90.83        645430.0  
18            Placed               8           95.83       1000399.0  
..               ...             ...             ...             ...  
487           Placed               7           55.00        797760.0  
489           Placed               7           60.83        667747.0  
490           Placed               6           56.67       1097955.0  
495           Placed               7           71.67        572544.0  
498           Placed               8           64.17        778646.0  

[128 rows x 18 columns]

In [10]:
#28. Filter students who are not placed and attended more than 75 sessions.
df[(df["placement_status"]=="not placed") & (df.sessions_attended>=75)]

Empty DataFrame
Columns: [student_id, student_name, gender, city, enrollment_date, course, batch_type, total_sessions, sessions_attended, avg_quiz_score, assignment_score, project_score, final_rating, learning_hours, placement_status, feedback_score, attendance_pct, salary_offered]
Index: []

In [11]:
#29. Filter students belonging to Pune, Bangalore, or Hyderabad.
df[df["city"].isin(["Pune","Bangalore","Hyderbad"])]

student_id student_name  gender       city enrollment_date  \
2          1003    Student_3    Male       Pune      2024-12-05   
5          1006    Student_6  Female       Pune      2024-09-18   
18         1019   Student_19  Female       Pune      2024-01-08   
19         1020   Student_20    Male       Pune      2024-12-25   
20         1021   Student_21  Female       Pune      2024-07-02   
..          ...          ...     ...        ...             ...   
490        1491  Student_491  Female  Bangalore      2024-08-09   
494        1495  Student_495    Male       Pune      2024-12-07   
497        1498  Student_498  Female  Bangalore      2024-03-21   
499        1500  Student_500  Female       Pune      2024-06-06   
502        1210  Student_210    Male  Bangalore      2024-04-04   

             course batch_type  total_sessions  sessions_attended  \
2      Data Science    Weekday             120                 73   
5     Generative AI    Weekend             120                 79   
18     Data Science    Weekend             120                115   
19     Data Science    Weekend             120                106   
20          AI & ML    Weekday             120                102   
..              ...        ...             ...                ...   
490    Data Science    Weekday             120                 68   
494         AI & ML    Weekend             120                 61   
497  Data Analytics    Weekday             120                 94   
499         AI & ML    Weekday             120                 73   
502         AI & ML    Weekday             120                 75   

     avg_quiz_score  assignment_score  project_score  final_rating  \
2                49                72             73           3.6   
5                64                98             57           2.6   
18               74                58             78           2.9   
19               80                55             71           2.9   
20               45                83             89           3.7   
..              ...               ...            ...           ...   
490              58                64             96           3.4   
494              69                77             52           4.1   
497              94                53             63           3.4   
499              57                79             78           3.8   
502              40                84             77           2.4   

     learning_hours placement_status  feedback_score  attendance_pct  \
2                 8           Placed               5           60.83   
5                30           Placed               6           65.83   
18               11           Placed               8           95.83   
19               13           Placed               7           88.33   
20               23       Not Placed               6           85.00   
..              ...              ...             ...             ...   
490              35           Placed               6           56.67   
494               9           Placed               8           50.83   
497              39           Placed               7           78.33   
499              21           Placed               7           60.83   
502              19           Placed               5           62.50   

     salary_offered  
2          501904.0  
5          305730.0  
18        1000399.0  
19         820610.0  
20              NaN  
..              ...  
490       1097955.0  
494        867354.0  
497        333896.0  
499       1115331.0  
502        750043.0  

[169 rows x 18 columns]

In [12]:
#30. Filter students enrolled in AI & ML or Generative AI courses.
df[df["course"].isin(["AI & ML","Generative AI"])]

student_id student_name  gender       city enrollment_date  \
0          1001    Student_1    Male  Hyderabad      2024-11-25   
1          1002    Student_2  Female  Hyderabad      2024-09-29   
3          1004    Student_4    Male      Delhi      2024-03-09   
4          1005    Student_5    Male     Mumbai      2024-02-03   
5          1006    Student_6  Female       Pune      2024-09-18   
..          ...          ...     ...        ...             ...   
500        1452  Student_452  Female     Mumbai      2024-03-07   
501        1276  Student_276  Female  Hyderabad      2024-10-15   
502        1210  Student_210    Male  Bangalore      2024-04-04   
503        1178  Student_178    Male      Delhi      2024-01-20   
504        1397  Student_397  Female     Mumbai      2024-05-27   

            course batch_type  total_sessions  sessions_attended  \
0    Generative AI    Weekday             120                108   
1          AI & ML    Weekday             120                110   
3          AI & ML    Weekend             120                 71   
4    Generative AI    Weekday             120                101   
5    Generative AI    Weekend             120                 79   
..             ...        ...             ...                ...   
500  Generative AI    Weekday             120                 68   
501  Generative AI    Weekday             120                 95   
502        AI & ML    Weekday             120                 75   
503  Generative AI    Weekday             120                 80   
504  Generative AI    Weekend             120                101   

     avg_quiz_score  assignment_score  project_score  final_rating  \
0                59                96             66           3.7   
1                57                93             83           4.5   
3                64                91             81           2.7   
4                41                55             50           2.1   
5                64                98             57           2.6   
..              ...               ...            ...           ...   
500              84                84             58           2.2   
501              92                81             91           3.2   
502              40                84             77           2.4   
503              97                92             66           4.7   
504              66                83             66           3.6   

     learning_hours placement_status  feedback_score  attendance_pct  \
0                 6           Placed               7           90.00   
1                33       Not Placed               8           91.67   
3                 5       Not Placed               8           59.17   
4                23           Placed               9           84.17   
5                30           Placed               6           65.83   
..              ...              ...             ...             ...   
500              19       Not Placed               8           56.67   
501              16       Not Placed               8           79.17   
502              19           Placed               5           62.50   
503              33           Placed               5           66.67   
504              25           Placed               7           84.17   

     salary_offered  
0          738963.0  
1               NaN  
3               NaN  
4          970105.0  
5          305730.0  
..              ...  
500             NaN  
501             NaN  
502        750043.0  
503        890611.0  
504       1003022.0  

[234 rows x 18 columns]

In [13]:
#31. Filter students whose average quiz score lies between 60 and 80.
df[df["avg_quiz_score"].between(60,80)]

student_id student_name  gender     city enrollment_date          course  \
3          1004    Student_4    Male    Delhi      2024-03-09         AI & ML   
5          1006    Student_6  Female     Pune      2024-09-18   Generative AI   
7          1008    Student_8    Male  Chennai      2024-07-23    Data Science   
8          1009    Student_9    Male   Mumbai      2024-05-05         AI & ML   
9          1010   Student_10  Female    Delhi      2024-08-22  Data Analytics   
..          ...          ...     ...      ...             ...             ...   
489        1490  Student_490    Male     Pune      2024-09-12    Data Science   
493        1494  Student_494  Female    Delhi      2024-12-20   Generative AI   
494        1495  Student_495    Male     Pune      2024-12-07         AI & ML   
498        1499  Student_499    Male   Mumbai      2024-07-20    Data Science   
504        1397  Student_397  Female   Mumbai      2024-05-27   Generative AI   

    batch_type  total_sessions  sessions_attended  avg_quiz_score  \
3      Weekend             120                 71              64   
5      Weekend             120                 79              64   
7      Weekend             120                109              72   
8      Weekend             120                 83              75   
9      Weekday             120                118              70   
..         ...             ...                ...             ...   
489    Weekend             120                 73              75   
493    Weekday             120                 94              75   
494    Weekend             120                 61              69   
498    Weekday             120                 77              72   
504    Weekend             120                101              66   

     assignment_score  project_score  final_rating  learning_hours  \
3                  91             81           2.7               5   
5                  98             57           2.6              30   
7                  77             72           4.3              14   
8                  61             96           2.6              19   
9                  50             89           4.2              26   
..                ...            ...           ...             ...   
489                92             62           4.1               6   
493                88             88           3.9              36   
494                77             52           4.1               9   
498                48             76           NaN              27   
504                83             66           3.6              25   

    placement_status  feedback_score  attendance_pct  salary_offered  
3         Not Placed               8           59.17             NaN  
5             Placed               6           65.83        305730.0  
7             Placed               6           90.83       1053790.0  
8             Placed               8           69.17        502210.0  
9         Not Placed               8           98.33             NaN  
..               ...             ...             ...             ...  
489           Placed               7           60.83        667747.0  
493           Placed               8           78.33        596618.0  
494           Placed               8           50.83        867354.0  
498           Placed               8           64.17        778646.0  
504           Placed               7           84.17       1003022.0  

[194 rows x 18 columns]

In [14]:
#32. Filter students whose assignment score lies between 70 and 90.
df[df["assignment_score"].between(70,90)]

student_id student_name  gender       city enrollment_date  \
2          1003    Student_3    Male       Pune      2024-12-05   
6          1007    Student_7    Male     Mumbai      2024-02-22   
7          1008    Student_8    Male    Chennai      2024-07-23   
14         1015   Student_15  Female      Delhi      2024-05-30   
20         1021   Student_21  Female       Pune      2024-07-02   
..          ...          ...     ...        ...             ...   
499        1500  Student_500  Female       Pune      2024-06-06   
500        1452  Student_452  Female     Mumbai      2024-03-07   
501        1276  Student_276  Female  Hyderabad      2024-10-15   
502        1210  Student_210    Male  Bangalore      2024-04-04   
504        1397  Student_397  Female     Mumbai      2024-05-27   

             course batch_type  total_sessions  sessions_attended  \
2      Data Science    Weekday             120                 73   
6    Data Analytics    Weekend             120                112   
7      Data Science    Weekend             120                109   
14   Data Analytics    Weekend             120                 88   
20          AI & ML    Weekday             120                102   
..              ...        ...             ...                ...   
499         AI & ML    Weekday             120                 73   
500   Generative AI    Weekday             120                 68   
501   Generative AI    Weekday             120                 95   
502         AI & ML    Weekday             120                 75   
504   Generative AI    Weekend             120                101   

     avg_quiz_score  assignment_score  project_score  final_rating  \
2                49                72             73           3.6   
6                93                89             94           3.2   
7                72                77             72           4.3   
14               93                71             64           3.9   
20               45                83             89           3.7   
..              ...               ...            ...           ...   
499              57                79             78           3.8   
500              84                84             58           2.2   
501              92                81             91           3.2   
502              40                84             77           2.4   
504              66                83             66           3.6   

     learning_hours placement_status  feedback_score  attendance_pct  \
2                 8           Placed               5           60.83   
6                29           Placed               6           93.33   
7                14           Placed               6           90.83   
14               29           Placed               7           73.33   
20               23       Not Placed               6           85.00   
..              ...              ...             ...             ...   
499              21           Placed               7           60.83   
500              19       Not Placed               8           56.67   
501              16       Not Placed               8           79.17   
502              19           Placed               5           62.50   
504              25           Placed               7           84.17   

     salary_offered  
2          501904.0  
6          707137.0  
7         1053790.0  
14        1128411.0  
20              NaN  
..              ...  
499       1115331.0  
500             NaN  
501             NaN  
502        750043.0  
504       1003022.0  

[189 rows x 18 columns]

In [15]:
#33. Filter students whose learning hours lie between 10 and 25.
df[df["learning_hours"].between(10,25)]

student_id student_name  gender       city enrollment_date  \
4          1005    Student_5    Male     Mumbai      2024-02-03   
7          1008    Student_8    Male    Chennai      2024-07-23   
8          1009    Student_9    Male     Mumbai      2024-05-05   
10         1011   Student_11    Male    Chennai      2024-02-12   
18         1019   Student_19  Female       Pune      2024-01-08   
..          ...          ...     ...        ...             ...   
499        1500  Student_500  Female       Pune      2024-06-06   
500        1452  Student_452  Female     Mumbai      2024-03-07   
501        1276  Student_276  Female  Hyderabad      2024-10-15   
502        1210  Student_210    Male  Bangalore      2024-04-04   
504        1397  Student_397  Female     Mumbai      2024-05-27   

            course batch_type  total_sessions  sessions_attended  \
4    Generative AI    Weekday             120                101   
7     Data Science    Weekend             120                109   
8          AI & ML    Weekend             120                 83   
10    Data Science    Weekday             120                102   
18    Data Science    Weekend             120                115   
..             ...        ...             ...                ...   
499        AI & ML    Weekday             120                 73   
500  Generative AI    Weekday             120                 68   
501  Generative AI    Weekday             120                 95   
502        AI & ML    Weekday             120                 75   
504  Generative AI    Weekend             120                101   

     avg_quiz_score  assignment_score  project_score  final_rating  \
4                41                55             50           2.1   
7                72                77             72           4.3   
8                75                61             96           2.6   
10               53                48             61           4.6   
18               74                58             78           2.9   
..              ...               ...            ...           ...   
499              57                79             78           3.8   
500              84                84             58           2.2   
501              92                81             91           3.2   
502              40                84             77           2.4   
504              66                83             66           3.6   

     learning_hours placement_status  feedback_score  attendance_pct  \
4                23           Placed               9           84.17   
7                14           Placed               6           90.83   
8                19           Placed               8           69.17   
10               25       Not Placed               7           85.00   
18               11           Placed               8           95.83   
..              ...              ...             ...             ...   
499              21           Placed               7           60.83   
500              19       Not Placed               8           56.67   
501              16       Not Placed               8           79.17   
502              19           Placed               5           62.50   
504              25           Placed               7           84.17   

     salary_offered  
4          970105.0  
7         1053790.0  
8          502210.0  
10              NaN  
18        1000399.0  
..              ...  
499       1115331.0  
500             NaN  
501             NaN  
502        750043.0  
504       1003022.0  

[228 rows x 18 columns]

In [ ]:
#34. Create a new column total_score as the sum of quiz, assignment, and project scores.

In [18]:
#35. Create a new column attendance_flag based on attendance percentage.
def percentage(x):
    if x>75:
        return "Good"
    else:
        return "Not Good"
df["attendance_flag"]=df["attendance_pct"].apply(percentage)

In [19]:
df.head()

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  \
0    Weekday             120                108              59   
1    Weekday             120                110              57   
2    Weekday             120                 73              49   
3    Weekend             120                 71              64   
4    Weekday             120                101              41   

   assignment_score  project_score  final_rating learning_hours  \
0                96             66           3.7            low   
1                93             83           4.5            low   
2                72             73           3.6            low   
3                91             81           2.7            low   
4                55             50           2.1            low   

  placement_status  feedback_score  attendance_pct  salary_offered  \
0           Placed               7           90.00        738963.0   
1       Not Placed               8           91.67             NaN   
2           Placed               5           60.83        501904.0   
3       Not Placed               8           59.17             NaN   
4           Placed               9           84.17        970105.0   

  attendance_flag  
0            Good  
1            Good  
2        Not Good  
3        Not Good  
4            Good

In [16]:
#36. Create a new column learning_intensity based on learning hours.
def learning_intensity(x):
    if x>=75:
        return "high"
    else:
        return "low"
df["learning_hours"]=df["learning_hours"].apply(learning_intensity)

In [17]:
df.head()

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  \
0    Weekday             120                108              59   
1    Weekday             120                110              57   
2    Weekday             120                 73              49   
3    Weekend             120                 71              64   
4    Weekday             120                101              41   

   assignment_score  project_score  final_rating learning_hours  \
0                96             66           3.7            low   
1                93             83           4.5            low   
2                72             73           3.6            low   
3                91             81           2.7            low   
4                55             50           2.1            low   

  placement_status  feedback_score  attendance_pct  salary_offered  
0           Placed               7           90.00        738963.0  
1       Not Placed               8           91.67             NaN  
2           Placed               5           60.83        501904.0  
3       Not Placed               8           59.17             NaN  
4           Placed               9           84.17        970105.0

In [42]:
#37. Classify students based on total_score using apply().


In [43]:
#38. Create a placement eligibility column using apply().

In [20]:
#39. Map placement_status values to numerical values.
df['placement_status'] = df['placement_status'].map({'Placed':1, 'Not Placed':0})

In [21]:
df.head()

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  \
0    Weekday             120                108              59   
1    Weekday             120                110              57   
2    Weekday             120                 73              49   
3    Weekend             120                 71              64   
4    Weekday             120                101              41   

   assignment_score  project_score  final_rating learning_hours  \
0                96             66           3.7            low   
1                93             83           4.5            low   
2                72             73           3.6            low   
3                91             81           2.7            low   
4                55             50           2.1            low   

   placement_status  feedback_score  attendance_pct  salary_offered  \
0                 1               7           90.00        738963.0   
1                 0               8           91.67             NaN   
2                 1               5           60.83        501904.0   
3                 0               8           59.17             NaN   
4                 1               9           84.17        970105.0   

  attendance_flag  
0            Good  
1            Good  
2        Not Good  
3        Not Good  
4            Good

In [ ]:
#40. Map batch_type values to abbreviated codes.
df['batch_type_code']= df['batch_type'].map(batch_type_map)

In [21]:
# 41.Map gender values to single-letter codes.
gender_val={
    "Female":"F",
    "Male":"M"}
df["gender"]=df["student_name"].map(gender_val)

In [22]:
df.head()

student_id student_name gender       city enrollment_date         course  \
0        1001    Student_1    NaN  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2    NaN  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    NaN       Pune      2024-12-05   Data Science   
3        1004    Student_4    NaN      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    NaN     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  \
0    Weekday             120                108              59   
1    Weekday             120                110              57   
2    Weekday             120                 73              49   
3    Weekend             120                 71              64   
4    Weekday             120                101              41   

   assignment_score  project_score  final_rating  learning_hours  \
0                96             66           3.7               6   
1                93             83           4.5              33   
2                72             73           3.6               8   
3                91             81           2.7               5   
4                55             50           2.1              23   

  placement_status  feedback_score  attendance_pct  salary_offered  
0           Placed               7           90.00        738963.0  
1       Not Placed               8           91.67             NaN  
2           Placed               5           60.83        501904.0  
3       Not Placed               8           59.17             NaN  
4           Placed               9           84.17        970105.0

In [5]:
#42.Count students under each learning_intensity category

In [25]:
#43.Calculate the percentage of placement-eligible students.
df["placement_status"].value_counts()%100

placement_status
1    36
0    69
Name: count, dtype: int64

In [32]:
#44.Identify the city with the highest number of placed stude
df["city"].max()

'Pune'

In [44]:
#45.Compare average total_score for placed and not placed students.

In [4]:
#46.Identify the top 5 students based on total_score.


KeyError: 'total_score'